src/embeddings.rs

```diff
@@ -415,6 +449,59 @@ where
     })
 }
 
+pub enum PyEmbeddingDefault {
+    Embedding(Py<PyArray1<f32>>),
+    Constant(f32),
+    None,
+}
+
+impl<'a> Default for PyEmbeddingDefault {
+    fn default() -> Self {
+        PyEmbeddingDefault::None
+    }
+}
+
+impl<'a> FromPyObject<'a> for PyEmbeddingDefault {
+    fn extract(ob: &'a PyAny) -> Result<Self, PyErr> {
+        if ob.is_none() {
+            return Ok(PyEmbeddingDefault::None);
+        }
+        let maybe_emb = ob
+            .extract()
+            .map(|e: &PyArray1<f32>| PyEmbeddingDefault::Embedding(e.to_owned()));
+        if maybe_emb.is_ok() {
+            return maybe_emb;
+        }
+        let maybe_const = ob.extract().map(PyEmbeddingDefault::Constant);
+        if maybe_const.is_ok() {
+            return maybe_const;
+        }
+        let maybe_iter = ob.iter();
+        if let Ok(iter) = maybe_iter {
+            let mut embed_vec = ob.len().map(Vec::with_capacity).unwrap_or_default();
+            let mut ok = true;
+            for item in iter {
+                let item = item.and_then(|item| item.extract());
+                match item {
+                    Ok(val) => embed_vec.push(val),
+                    Err(_) => {
+                        ok = false;
```

Comment:
Return with an error immediately?

Reply:
Sure, that's a remnant of when the iter-check was not the final block.